### PR TITLE
Linter support

### PR DESCRIPTION
**Pull request by dvcrn**
Refs https://github.com/dvcrn/proton/issues/50.

We need to check how we could deal with linters that automatically install `Linter`. In particular the [clojure linter](https://github.com/AtomLinter/linter-clojure) does this, but solving this problem will be very helpful later on

### CODE REVIEW

**Comment by dvcrn** [comment]
@geksilla ok I experimented a bit with this but ran into a dead end:

I used your code to create a `layer-deps` multi that does the same thing that your method did, just for layers. The problem is that multimethods can only get defined once:

```
;; clojure
(defmethod layer-deps :tools/linter []
 [:linter-clojure])
```

```
;; elixir
(defmethod layer-deps :tools/linter []
 [:linter-elixirc])
```

`:linter-elixirc` will overwrite `:linter-clojure`. We need something else for this.

**Comment by dvcrn** [comment]
All right, I went with this approach:

```
(register-layer-dependencies :tools/linter 
    [:linter-clojure])
```

I think this is ready to review. Ping @geksilla @spencerlyon2

File: src/cljs/proton/core.cljs [hunk 1]
```diff
@@ -108,9 +109,7 @@
 
           ;; wipe existing config
           (atom-env/insert-process-step! "Wiping existing configuration")
-          (doall (map atom-env/unset-config! (filter #(not
-                                                        (or (= "core.disabledPackages" %)
-                                                            (= "core.themes" %))) (atom-env/get-all-settings))))
+          (doall (map atom-env/unset-config! (filter #(not (= "core.themes" %)) (atom-env/get-all-settings))))
```
Comment (spencerlyon2) [inline src/cljs/proton/core.cljs]: Why the change with `core.disabledPackages`?

Comment (dvcrn) [inline src/cljs/proton/core.cljs]: We had a faulty state when packages were disabled / uninstalled. So even if a package gets uninstalled and installed again, it is still deactivated and proton currently has no way of activating one. 

This conflicted with the uninstallation process of, for example, linters. Apparently the atom API doesn't return deactivated packages as `installed` (at least in my tests) so we couldn't purge them. They would just remain as disabled zombie packages.

I tested this a bit and it seems like the problem we initially tried to solve with this (packages getting activated right away after they are installed through proton) is not present anymore. Though we need to test this a bit more :)

File: src/cljs/proton/layers/base.cljs [hunk 2]
```diff
@@ -10,3 +12,43 @@
 (defmulti get-keybindings dispatch)
 (defmulti get-keymaps dispatch)
 (defmulti describe-mode dispatch)
+
+(defn register-package-dependencies [key deps]
```
Comment (spencerlyon2) [inline src/cljs/proton/layers/base.cljs]: What's the distinction between `register-package-dependencies` and `register-layer-dependencies`.

It seems like we are only using the latter. What is the intended purpose for `register-package-dependencies`?

Comment (dvcrn) [inline src/cljs/proton/layers/base.cljs]: That was initially a leftover from a first version. I tried to tie the linter packages to the linter package itself. 

I think we could remove it

Comment (spencerlyon2) [inline src/cljs/proton/layers/base.cljs]: I think it makes sense to tie layer dependencies to a complete layer, not just a package. I'm in favor of removing

File: src/cljs/proton/core.cljs (hunk 1 above)
Comment (dvcrn) [inline src/cljs/proton/core.cljs]: You were right with this one. This messes with layers that enable/disable packages on init. Need to put this back in.

**Comment by spencerlyon2** [comment]
ping. Are we ready to go here?

**Comment by dvcrn** [comment]
A bit low on time lately I'll try to finish it today but no promises
